crates/turborepo-lib/src/rewrite_json.rs

```diff
@@ -0,0 +1,411 @@
+ * encountered with the input json_document_string.
+ */
+fn get_root(json_document_string: &str) -> Result<jsonc_parser::ast::Value, RewriteError> {
+    let parse_result_result = parse_to_ast(
+        json_document_string,
+        &Default::default(),
+        &Default::default(),
+    );
+
+    return match parse_result_result {
+        Ok(parse_result) => match parse_result.value {
+            Some(root) => Ok(root),
+            None => Err(RewriteError::NoRoot),
+        },
+        Err(parse_error) => Err(RewriteError::ParseError(parse_error)),
+    };
+}
+
+/**
+ * get_closest_node traverses the JSON document via recursive tail calls to
+ * find the last node that exists with in this JSON key path.
+ */
+fn get_closest_node<'a>(
+    current_node: &'a jsonc_parser::ast::Value<'a>,
+    target_path: &Vec<&'a str>,
+    current_path: Vec<&'a str>,
+) -> (Vec<&'a str>, &'a jsonc_parser::ast::Value<'a>) {
+    // No target_path? We've arrived.
+    if target_path.len() == 0 {
+        return (current_path, current_node);
+    }
+
+    match current_node {
+        // Only objects can have key paths.
+        jsonc_parser::ast::Value::Object(obj) => {
+            let object_property = obj.properties.iter().rev().find(|property| {
+                let current_property_name = property.name.as_str();
+                return target_path[0] == current_property_name;
+            });
+
+            // See if we found a matching key. If so, recurse.
+            match object_property {
+                Some(property) => {
+                    let next_current_node = &property.value;
+                    let next_property_name = property.name.as_str();
+                    let mut next_current_path = current_path.clone();
+                    next_current_path.push(next_property_name);
+                    let next_target_path = &target_path[1..].to_vec();
+
+                    // Tail call!
+                    get_closest_node(next_current_node, next_target_path, next_current_path)
+                }
+                None => (current_path, current_node),
+            }
+        }
+        // All other node types are complete.
+        _ => (current_path, current_node),
+    }
+}
+
+/**
+ * Given path segments, generate a JSON object.
+ */
+fn generate_object(path_segments: Vec<&str>, value: &str) -> String {
+    let mut output = String::new();
+    let length = path_segments.len();
+
+    for path in path_segments {
+        output.push_str("{");
+        output.push_str("\"");
+        output.push_str(path);
+        output.push_str("\"");
+        output.push_str(":");
+    }
+    output.push_str(value);
+    for _ in 0..length {
+        output.push_str("}");
+    }
+
+    return output;
+}
+
+/**
+ * Given path segments, generate a JSON object member with an optional
+ * trailing separator.
+ */
+fn generate_path(path_segments: Vec<&str>, value: &str, separator: &str) -> String {
```

Comment:
Do we escape stuff like `"` when writing back the JSON?

Reply:
Nope, I tried to make that clear in the comments:

> Given a JSONC document, an object traversal path, and a _pre-serialized_ JSON value `set_path` will return a minimally-mutated JSONC document...

You'll also see that I format the spaceid string in `link.rs`. I consider this out of scope mostly because YAGNI. If we actually do need it I'm going to be finishing a PR to Rome.